### PR TITLE
Add Firewalld AutoYaST support (fate#323460)

### PR DESCRIPTION
**Pull request by teclator**
- [Trello Card](https://trello.com/c/jL8fhEKw/1091-5-firewall-firewalld-is-the-new-susefirewall-autoyast-support)

### CODE REVIEW

File: src/lib/y2firewall/importer_strategies/suse_firewall.rb [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+          zone = firewalld.find_zone(zone.equivalent(ipsec_trust_zone))
+          zone.services << "ipsec"
+        end
+
+        firewalld.log_denied_packets = log_denied_packets
+
+        true
+      end
+
+      # Given a SuSEFirewall2 zone name it process the profile's configuration
```
Comment (imobachgs) [inline src/lib/y2firewall/importer_strategies/suse_firewall.rb]: *it processes*

File: src/lib/y2firewall/importer_strategies/suse_firewall.rb [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+      # Constructor
+      #
+      # @param [Hash] AutoYaST profile firewall's section
+      def initialize(profile)
+        @profile = profile
+      end
+
+      # It process the profile configuring the firewalld zones that match
+      # better with the SuSEFirewalld ones.
```
Comment (imobachgs) [inline src/lib/y2firewall/importer_strategies/suse_firewall.rb]: SuSEFirewalld?

File: src/lib/y2firewall/importer_strategies/suse_firewall.rb [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    private
+
+      # Obtain the IP ports for the given SuSEFIrewall2 zone name from the
+      # profile.
+      #
+      # @param zone_name [String]
+      # @return [Array<Strint>, nil] list of configured IPP ports; nil if no
+      # configured
+      def ip_ports(zone)
+        ports = profile["FW_SERVICES_#{zone}_IP"]
```
Comment (imobachgs) [inline src/lib/y2firewall/importer_strategies/suse_firewall.rb]: Maybe you would save this value in an instance variable. It is not a big deal, but you are calling this method twice from `#ports` and I guess that you do not expect the profile to change.

File: src/lib/y2firewall/importer_strategies/suse_firewall.rb [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
```
Comment (imobachgs) [inline src/lib/y2firewall/importer_strategies/suse_firewall.rb]: really-np: why a blank line in a two-lines method?

File: src/lib/y2firewall/importer_strategies/suse_firewall.rb [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        if drop_all
+          "all"
+        elsif accept_crit || drop_crit
+          "unicast"
+        else
+          "none"
+        end
+      end
+
+      def firewalld
```
Comment (imobachgs) [inline src/lib/y2firewall/importer_strategies/suse_firewall.rb]: Same about memoization

File: test/lib/y2firewall/clients/auto_test.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  let(:importer) { double("Y2Firewall::Importer", import: true) }
+
+  before do
+    allow(firewalld).to receive(:read)
+    allow(subject).to receive(:importer).and_return(importer)
+  end
+
+  describe "#summary" do
+    it "returns the summary of all the configured zones" do
+      expect(firewalld.api).to receive(:list_all_zones).and_return(["zone1", "zone2"])
```
Comment (imobachgs) [inline test/lib/y2firewall/clients/auto_test.rb]: I would say that it is a "precondition", but not an expectation at all. I would switch it to `allow` (actually I would move it to a before block, but maybe that's too much).

File: test/lib/y2firewall/clients/auto_test.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    let(:arguments) { { "FW_MASQUERADE" => "yes" } }
+
+    it "reads the current firewalld configuration" do
+      expect(firewalld).to receive(:read)
+
+      subject.import(arguments)
+    end
+
+    it "pass its arguments to the firewalld importer" do
+      expect(subject).to receive(:importer).and_return(importer)
```
Comment (imobachgs) [inline test/lib/y2firewall/clients/auto_test.rb]: Expectation?

File: test/lib/y2firewall/importer_strategies/suse_firewall.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+#
+# To contact SUSE about this file by physical or electronic mail, you may find
+# current contact information at www.suse.com.
+# ------------------------------------------------------------------------------
+
+require_relative "../../../test_helper.rb"
+require "cwm/rspec"
+require "y2firewall/importer_strategies/suse_firewall"
+
+describe Y2Firewall::ImporterStrategies::SuseFirewall do
```
Comment (imobachgs) [inline test/lib/y2firewall/importer_strategies/suse_firewall.rb]: Missing tests.

File: test/lib/y2firewall/importer_test.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+require_relative "../../test_helper.rb"
+require "y2firewall/importer"
+
+describe Y2Firewall::Importer do
+  let(:profile) { { "FW_DEV_EXT" => "eth0" } }
+
+  describe "#import" do
+    context "when the given profile uses a SuSEFirewall2 schema" do
+      it "imports the given profile using the SuSEFirewall strategy" do
+        expect(subject).to receive(:strategy_for).with(profile).and_call_original
```
Comment (imobachgs) [inline test/lib/y2firewall/importer_test.rb]: expectation?

File: src/lib/y2firewall/importer.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  # This class is responsible of import firewalld AutoYaST configuration
+  # supporting the new firewalld schema but also the SuSEFirewall one.
+  class Importer
+    # Import the given configuration
+    #
+    # @param [Hash] AutoYaST profile firewall's section
+    # @return [true,nil] return true if success; return nil if the given
+    # profile is empty
+    def import(profile)
+      return if profile.empty?
```
[the remark this answers is not in the document]
Comment (teclator) [reply src/lib/y2firewall/importer.rb]: In that case the importer should not be called, it is not expected to be nil

File: src/lib/y2firewall/importer_strategies/suse_firewall.rb (hunk 2 above)
Comment (teclator) [reply src/lib/y2firewall/importer_strategies/suse_firewall.rb]: xDDD too many names .. I meant SuSEFirewall2 fixing it.

**Pull request merged**